<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="fr" xml:lang="fr"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Arthur Oulès, adapté de Güell et al (2017)">

<title>TD Biodiversité</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="biodiversity_tutorial_files/libs/clipboard/clipboard.min.js"></script>
<script src="biodiversity_tutorial_files/libs/quarto-html/quarto.js"></script>
<script src="biodiversity_tutorial_files/libs/quarto-html/popper.min.js"></script>
<script src="biodiversity_tutorial_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="biodiversity_tutorial_files/libs/quarto-html/anchor.min.js"></script>
<link href="biodiversity_tutorial_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="biodiversity_tutorial_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="biodiversity_tutorial_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="biodiversity_tutorial_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="biodiversity_tutorial_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article toc-left">
<div id="quarto-sidebar-toc-left" class="sidebar toc-left">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Sommaire</h2>
   
  <ul>
  <li><a href="#chargement-des-packages-et-données" id="toc-chargement-des-packages-et-données" class="nav-link active" data-scroll-target="#chargement-des-packages-et-données">Chargement des packages et données</a>
  <ul class="collapse">
  <li><a href="#chargement-des-packages" id="toc-chargement-des-packages" class="nav-link" data-scroll-target="#chargement-des-packages">Chargement des packages</a></li>
  <li><a href="#chargement-des-fonctions-externes" id="toc-chargement-des-fonctions-externes" class="nav-link" data-scroll-target="#chargement-des-fonctions-externes">Chargement des fonctions externes</a></li>
  </ul></li>
  <li><a href="#préliminaires" id="toc-préliminaires" class="nav-link" data-scroll-target="#préliminaires">Préliminaires</a></li>
  <li><a href="#acquisition-des-séquences" id="toc-acquisition-des-séquences" class="nav-link" data-scroll-target="#acquisition-des-séquences">Acquisition des séquences</a>
  <ul class="collapse">
  <li><a href="#chargement-des-données" id="toc-chargement-des-données" class="nav-link" data-scroll-target="#chargement-des-données">Chargement des données</a></li>
  <li><a href="#récupération-des-séquences-par-lot" id="toc-récupération-des-séquences-par-lot" class="nav-link" data-scroll-target="#récupération-des-séquences-par-lot">Récupération des séquences par lot</a></li>
  </ul></li>
  <li><a href="#alignement-des-séquences" id="toc-alignement-des-séquences" class="nav-link" data-scroll-target="#alignement-des-séquences">Alignement des séquences</a></li>
  <li><a href="#méthode-de-distances" id="toc-méthode-de-distances" class="nav-link" data-scroll-target="#méthode-de-distances">Méthode de distances</a>
  <ul class="collapse">
  <li><a href="#génération-de-la-matrice-de-distances" id="toc-génération-de-la-matrice-de-distances" class="nav-link" data-scroll-target="#génération-de-la-matrice-de-distances">Génération de la matrice de distances</a></li>
  <li><a href="#génération-et-visualisation-de-larbre-par-upgma" id="toc-génération-et-visualisation-de-larbre-par-upgma" class="nav-link" data-scroll-target="#génération-et-visualisation-de-larbre-par-upgma">Génération et visualisation de l’arbre par UPGMA</a></li>
  <li><a href="#génération-de-larbre-par-neighbour-joining" id="toc-génération-de-larbre-par-neighbour-joining" class="nav-link" data-scroll-target="#génération-de-larbre-par-neighbour-joining">Génération de l’arbre par Neighbour-Joining</a></li>
  <li><a href="#quelle-méthode-choisir" id="toc-quelle-méthode-choisir" class="nav-link" data-scroll-target="#quelle-méthode-choisir">Quelle méthode choisir ?</a></li>
  <li><a href="#bootstrap" id="toc-bootstrap" class="nav-link" data-scroll-target="#bootstrap">Bootstrap</a>
  <ul class="collapse">
  <li><a href="#calcul-des-valeurs-de-bootstrap" id="toc-calcul-des-valeurs-de-bootstrap" class="nav-link" data-scroll-target="#calcul-des-valeurs-de-bootstrap">Calcul des valeurs de Bootstrap</a></li>
  <li><a href="#impression-de-larbre-bootstrappé" id="toc-impression-de-larbre-bootstrappé" class="nav-link" data-scroll-target="#impression-de-larbre-bootstrappé">Impression de l’arbre bootstrappé</a></li>
  <li><a href="#on-retire-les-noeuds-faibles" id="toc-on-retire-les-noeuds-faibles" class="nav-link" data-scroll-target="#on-retire-les-noeuds-faibles">On retire les noeuds faibles</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#maximum-de-parcimonie" id="toc-maximum-de-parcimonie" class="nav-link" data-scroll-target="#maximum-de-parcimonie">Maximum de parcimonie</a>
  <ul class="collapse">
  <li><a href="#conversion-de-lalignement-dans-le-format-phydat" id="toc-conversion-de-lalignement-dans-le-format-phydat" class="nav-link" data-scroll-target="#conversion-de-lalignement-dans-le-format-phydat">Conversion de l’alignement dans le format phyDat</a></li>
  <li><a href="#génération-de-larbre-initial" id="toc-génération-de-larbre-initial" class="nav-link" data-scroll-target="#génération-de-larbre-initial">Génération de l’arbre initial</a></li>
  <li><a href="#génération-de-larbre-le-plus-parcimonieux" id="toc-génération-de-larbre-le-plus-parcimonieux" class="nav-link" data-scroll-target="#génération-de-larbre-le-plus-parcimonieux">Génération de l’arbre le plus parcimonieux</a></li>
  </ul></li>
  <li><a href="#maximum-de-vraisemblance" id="toc-maximum-de-vraisemblance" class="nav-link" data-scroll-target="#maximum-de-vraisemblance">Maximum de vraisemblance</a>
  <ul class="collapse">
  <li><a href="#initialisation" id="toc-initialisation" class="nav-link" data-scroll-target="#initialisation">Initialisation</a>
  <ul class="collapse">
  <li><a href="#calcul-de-la-vraisemblance-de-larbre-initial" id="toc-calcul-de-la-vraisemblance-de-larbre-initial" class="nav-link" data-scroll-target="#calcul-de-la-vraisemblance-de-larbre-initial">Calcul de la vraisemblance de l’arbre initial</a></li>
  </ul></li>
  <li><a href="#génération-de-larbre-le-plus-vraisemblable" id="toc-génération-de-larbre-le-plus-vraisemblable" class="nav-link" data-scroll-target="#génération-de-larbre-le-plus-vraisemblable">Génération de l’arbre le plus vraisemblable</a></li>
  </ul></li>
  </ul>
</nav>
</div>
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
</div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title">TD Biodiversité</h1><button type="button" class="btn code-tools-button dropdown-toggle" id="quarto-code-tools-menu" data-bs-toggle="dropdown" aria-expanded="false"><i class="bi"></i> Code</button><ul class="dropdown-menu dropdown-menu-end" aria-labelelledby="quarto-code-tools-menu"><li><a id="quarto-show-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Montrer tout le code</a></li><li><a id="quarto-hide-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Cacher tout le code</a></li><li><hr class="dropdown-divider"></li><li><a id="quarto-view-source" class="dropdown-item" href="javascript:void(0)" role="button" data-quarto-source-url="https://github.com/Arthur-Oules/TD_Biodiversite/blob/main/biodiversity_tutorial.qmd">Voir les sources</a></li></ul></div></div>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Auteur·rice</div>
    <div class="quarto-title-meta-contents">
             <p>Arthur Oulès, adapté de <a href="https://fuzzyatelin.github.io/bioanth-stats/module-24/module-24.html#introduction">Güell <em>et al</em> (2017)</a> </p>
          </div>
  </div>
    
  
    
  </div>
  

</header>

<section id="chargement-des-packages-et-données" class="level1">
<h1>Chargement des packages et données</h1>
<section id="chargement-des-packages" class="level2">
<h2 class="anchored" data-anchor-id="chargement-des-packages">Chargement des packages</h2>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># For data manipulation</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(here)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(reshape2)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(rentrez)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="co"># For Phylogenetics</span></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ape)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(msa)</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(phangorn)</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="co"># For visualization</span></span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggtree)</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(viridis)</span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="chargement-des-fonctions-externes" class="level2">
<h2 class="anchored" data-anchor-id="chargement-des-fonctions-externes">Chargement des fonctions externes</h2>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="fu">here</span>(<span class="st">"functions"</span>, <span class="st">"tidy_functions.R"</span>))</span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
</section>
<section id="préliminaires" class="level1">
<h1>Préliminaires</h1>
<p>Dans le cadre de ce TP/TD, nous allons apprendre à générer des arbres phylogénétiques à partir de séquences ADN récupérées sur GenBank par diverses méthodes de construction d’arbre et les comparer entre elles.</p>
<p>Quel est le protocole général pour construire un arbre phylogénétique à partir de séquences ADN ?</p>
</section>
<section id="acquisition-des-séquences" class="level1">
<h1>Acquisition des séquences</h1>
<p>Nous allons commencer par récupérer les séquences d’intérêts sur Genbank directement depuis <code>R</code>.<br>
Qu’est-ce que GenBank ? (5 min)</p>
<section id="chargement-des-données" class="level2">
<h2 class="anchored" data-anchor-id="chargement-des-données">Chargement des données</h2>
<p>Charger le fichier <code>data/usflu.annot.csv</code> avec la fonction <code>read_csv()</code> et le visualiser. (5 min)</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>annotations <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">file           =</span> <span class="fu">here</span>(<span class="st">"data"</span>, <span class="st">"usflu.annot.csv"</span>),</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">show_col_types =</span> <span class="cn">FALSE</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>) <span class="sc">|&gt;</span> <span class="fu">select</span>(<span class="fu">c</span>(<span class="sc">-</span><span class="dv">1</span>))</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="fu">view</span>(annotations)</span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Aller sur <a href="https://www.ncbi.nlm.nih.gov/nucleotide/">GenBank</a> et entrer un numéro d’accès dans la barre de recherche. À quoi correspond-t-elle ?<br>
Pourquoi l’indication “Segment 4” ?<br>
À quoi correspond la protéine codée par ce gène ? Quel est le rapport avec le nom du variant ?<br>
Pouvez-vous faire des prédictions sur l’évolution de ces séquences ? (10 min)</p>
</section>
<section id="récupération-des-séquences-par-lot" class="level2">
<h2 class="anchored" data-anchor-id="récupération-des-séquences-par-lot">Récupération des séquences par lot</h2>
<p>Nous allons télécharger les séquences correspondantes depuis Genbank avec le package <code>rentrez</code> (ne pas faire tourner le bloc sauf avec une bonne connexion internet).</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>annotations <span class="sc">|&gt;</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="st">"sequences"</span> <span class="ot">=</span> <span class="fu">Vectorize</span>(Get_sequence)(accession)) <span class="sc">|&gt;</span> </span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="fu">c</span>(accession, sequences)) <span class="sc">|&gt;</span> </span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">write_to_fasta</span>(<span class="at">path =</span> <span class="fu">here</span>(<span class="st">"output"</span>, <span class="st">"usflu_sequences.fasta"</span>))</span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
</section>
<section id="alignement-des-séquences" class="level1">
<h1>Alignement des séquences</h1>
<p>Rappeler le principe de l’alignement de séquences. (5min)<br>
</p>
<p>Le package <code>msa</code> (Multiple Sequence Alignment) permet d’utiliser les algorithmes classiques d’alignement de multiples séquences directement depuis <code>R</code>.<br>
Ouvrir la page d’aide de la fonction <code>msa()</code>.<br>
Quels sont les algorithmes disponibles ? Sous quelle forme doit-on fournir les séquences à la fonction ? (5 min)<br>
Commencer par ouvrir le fichier fasta généré dans un format accepté par la fonction <code>msa()</code>.</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>usflu_sequences <span class="ot">&lt;-</span> <span class="fu">readDNAStringSet</span>(<span class="at">file =</span> <span class="fu">here</span>(<span class="st">"output"</span>, <span class="st">"usflu_sequences.fasta"</span>))</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>usflu_sequences</span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>DNAStringSet object of length 80:
     width seq                                              names               
 [1]  1701 ATGAAGACTATCATTGCTTTGAG...TAGGTGCAACATTTGCATTTGA CY013200
 [2]  1701 ATGAAGACTATCATTGCTTTGAG...TAGGTGCAACATTTGCATTTGA CY013781
 [3]  1701 ATGAAGACTATCATTGCTTTGAG...TAGGTGCAACATTTGCATTTGA CY012128
 [4]  1701 ATGAAGACTATCATTGCTTTGAG...TAGGTGCAACATTTGCATTTGA CY013613
 [5]  1701 ATGAAGACTATCATTGCTTTGAG...TAGGTGCAACATTTGCATTTGA CY012160
 ...   ... ...
[76]  1701 ATGAAGACTATCATTGCTTTGAG...TAGGTGCAACATTTGCATTTGA FJ549055
[77]  1701 ATGAAGACTATCATTGCTTTGAG...TAGGTGCAACATTTGCATTTGA EU779498
[78]  1701 ATGAAGACTATCATTGCTTTGAG...TAGATGCAACATTTGCATTTGA EU779500
[79]  1701 ATGAAGACTATCATTGCTTTGAG...TAGGTGCAACATTTGCATTTGA CY035190
[80]  1701 ATGAAGACTATCATTGCTTTGAG...TAGGTGCAACATTTGCATTTGA EU852005</code></pre>
</div>
</div>
<p>On va utiliser l’algorithme MUSCLE pour générer notre alignement.<br>
Utiliser la fonction <code>msa()</code> avec les bons paramètres pour générer un alignement. (5 min)</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>alignment <span class="ot">&lt;-</span> usflu_sequences <span class="sc">|&gt;</span> <span class="fu">msa</span>(<span class="at">method =</span> <span class="st">"Muscle"</span>, <span class="at">type =</span> <span class="st">"dna"</span>, <span class="at">order =</span> <span class="st">"input"</span>)</span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Ouvrir le fichier <code>output/alignment.pdf</code> et observer l’alignement.</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">system2</span>(</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>  <span class="st">"open"</span>,</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">args =</span> <span class="fu">c</span>(<span class="st">"-a Preview.app"</span>, <span class="fu">here</span>(<span class="st">"output"</span>, <span class="st">"alignment.pdf"</span>)),</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">wait =</span> <span class="cn">FALSE</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Maintenant que nous avons aligné nos séquences, nous pouvons passer à la suite !<br>
Rappeler les différentes méthodes de construction d’arbres phylogénétiques que vous connaissez. (5 min)</p>
</section>
<section id="méthode-de-distances" class="level1">
<h1>Méthode de distances</h1>
<p>Rappeler le principe des méthodes de distance et en citer deux exemples.<br>
Quelles sont les différences entre ces méthodes ? (5 min)</p>
<section id="génération-de-la-matrice-de-distances" class="level2">
<h2 class="anchored" data-anchor-id="génération-de-la-matrice-de-distances">Génération de la matrice de distances</h2>
<p>Nous allons générer une matrice des distances à partir de nos alignements à l’aide la fonction <code>dist.dna()</code>. Ouvrir l’aide de cette fonction.<br>
Combien de modèles de substitution sont disponibles ? Comparer à la fonction <code>dist.ml()</code>. Aller sur <a href="http://www.iqtree.org/doc/Substitution-Models">ce lien</a> (et <a href="https://en.wikipedia.org/wiki/Models_of_DNA_evolution">celui-ci</a> pour approfondir) et regarder les différents modèles.<br>
Quelles sont les différences entre ces modèles ? Nous allons utiliser le modèle <code>TN93</code>, pourquoi ce nom ? Quelles sont ses particularités ? (5 min)<br>
Utiliser la fonction <code>dist.dna()</code> pour générer une matrice de distance.</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># alignment_BIN &lt;- adegenet::fasta2DNAbin(here("output", "alignment.fa"))</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>alignment_BIN <span class="ot">&lt;-</span> alignment <span class="sc">|&gt;</span> <span class="fu">msaConvert</span>(<span class="at">type =</span> <span class="st">"ape::DNAbin"</span>) </span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>distance_matrix <span class="ot">&lt;-</span> alignment_BIN <span class="sc">|&gt;</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">dist.dna</span>(<span class="at">model =</span> <span class="st">"TN93"</span>)</span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Nous allons visualiser la matrice de distance avec ggplot. Décrire la structure de cette portion de code. (5 min)</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>distance_matrix <span class="sc">|&gt;</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.matrix</span>() <span class="sc">|&gt;</span> <span class="fu">melt</span>() <span class="sc">|&gt;</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="at">mapping =</span> <span class="fu">aes</span>(<span class="at">x =</span> Var1, <span class="at">y =</span> Var2, <span class="at">fill =</span> value)) <span class="sc">+</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_raster</span>() <span class="sc">+</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_viridis</span>() <span class="sc">+</span></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"Heatmap de la matrice des distances"</span>) <span class="sc">+</span></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(</span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">axis.title =</span> <span class="fu">element_blank</span>()</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="biodiversity_tutorial_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Pourquoi les valeurs en diagonale de la matrice sont nulles ? Identifiez-vous des tendances ? (5 min)</p>
</section>
<section id="génération-et-visualisation-de-larbre-par-upgma" class="level2">
<h2 class="anchored" data-anchor-id="génération-et-visualisation-de-larbre-par-upgma">Génération et visualisation de l’arbre par UPGMA</h2>
<p>Générer et visualiser un arbre avec l’algorithme UPGMA à l’aide des fonction <code>upgma()</code> et <code>plot()</code>. Commenter cet arbre. (5 min)</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>UPGMA_tree <span class="ot">&lt;-</span> <span class="fu">upgma</span>(distance_matrix)</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(UPGMA_tree, <span class="at">main =</span> <span class="st">"Arbre UPGMA"</span>, <span class="at">cex =</span> .<span class="dv">6</span>)</span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="biodiversity_tutorial_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid" width="672"></p>
</div>
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># ggtree(UPGMA_tree) + geom_tiplab() + ggtitle("Arbre UPGMA")</span></span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="génération-de-larbre-par-neighbour-joining" class="level2">
<h2 class="anchored" data-anchor-id="génération-de-larbre-par-neighbour-joining">Génération de l’arbre par Neighbour-Joining</h2>
<p>Ouvrir l’aide de la fonction <code>NJ()</code>. Générer l’arbre et le visualiser avec la fonction plot().</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>NJ_tree <span class="ot">&lt;-</span> distance_matrix <span class="sc">|&gt;</span> <span class="fu">NJ</span>()</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(NJ_tree)</span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="biodiversity_tutorial_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Nous allons en profiter pour explorer la structure des arbres phylogénétiques en tant qu’objet sur R.<br>
Pour commencer, explorer la structure de l’objet <code>NJ_tree</code> à l’aide de la fonction <code>class()</code> et de l’opérateur <code>$</code><br>
Lire la documentation de la fonction <code>read.tree</code>.<br>
Enregistrer l’arbre dans un fichier <code>.txt</code> à l’aide de la fonction <code>ape::write.tree()</code>. Comment est structuré ce fichier texte ?<br>
Utiliser les fonctions <code>plot()</code> et <code>nodelabel()</code> pour afficher l’arbre et les numéros des nœuds. (10 min)</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(NJ_tree)</span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "phylo"</code></pre>
</div>
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>NJ_tree</span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>
Phylogenetic tree with 80 tips and 78 internal nodes.

Tip labels:
  CY013200, CY013781, CY012128, CY013613, CY012160, CY012272, ...

Unrooted; includes branch lengths.</code></pre>
</div>
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>NJ_tree <span class="sc">|&gt;</span> ape<span class="sc">::</span><span class="fu">write.tree</span>(<span class="fu">here</span>(<span class="st">"output"</span>, <span class="st">"NJ_tree.txt"</span>))</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(NJ_tree, <span class="at">cex =</span> .<span class="dv">6</span>)</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a><span class="fu">nodelabels</span>(<span class="at">frame =</span> <span class="st">"none"</span>, <span class="at">cex =</span> .<span class="dv">6</span>)</span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="biodiversity_tutorial_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Nous allons voir comment visualiser ce même arbre en utilisant le package ggtree.<br>
Commenter le bloc de code suivant. Cette représentation vous semble-t-elle pertinente pour cette méthode ? (5 min)</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggtree</span>(NJ_tree) <span class="sc">+</span> <span class="fu">geom_tiplab</span>() <span class="sc">+</span> <span class="fu">geom_nodelab</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"NJ non raciné"</span>)</span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="biodiversity_tutorial_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Ajoutons des annotations à cet arbre :</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggtree</span>(NJ_tree, <span class="at">layout =</span> <span class="st">"circular"</span>) <span class="sc">%&lt;+%</span> annotations <span class="sc">+</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_tiplab</span>(<span class="fu">aes</span>(<span class="at">color =</span> year)) <span class="sc">+</span></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_viridis</span>(<span class="at">option =</span> <span class="st">"turbo"</span>) <span class="sc">+</span></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"NJ non raciné"</span>) <span class="sc">+</span></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"right"</span>)</span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="biodiversity_tutorial_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>A quoi correspondent les couleurs des feuilles de l’arbre ?<br>
Nous allons raciner l’arbre de manière arbitraire. À partir de la figure précédente, quelle séquence vous semble pertinente pour raciner l’arbre ?<br>
Raciner l’arbre à l’aide de la fonction <code>root()</code>. (5 min)<br>
</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>NJ_tree_rooted <span class="ot">&lt;-</span> NJ_tree <span class="sc">|&gt;</span> <span class="fu">root</span>(<span class="at">outgroup =</span> <span class="st">"CY012128"</span>) <span class="sc">|&gt;</span> <span class="fu">ladderize</span>()</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggtree</span>(NJ_tree_rooted) <span class="sc">%&lt;+%</span> annotations <span class="sc">+</span></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_tiplab</span>(<span class="fu">aes</span>(<span class="at">color =</span> year)) <span class="sc">+</span></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_viridis</span>(<span class="at">option =</span> <span class="st">"turbo"</span>) <span class="sc">+</span></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"NJ raciné"</span>) <span class="sc">+</span></span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"right"</span>) <span class="sc">+</span></span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_rootedge</span>(<span class="at">rootedge =</span> .<span class="dv">01</span>) <span class="sc">+</span></span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_tree2</span>()</span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>! The tree contained negative edge lengths. If you want to ignore the edges,
you can set `options(ignore.negative.edge=TRUE)`, then re-run ggtree.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="biodiversity_tutorial_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Commenter cet arbre. (5 min)</p>
</section>
<section id="quelle-méthode-choisir" class="level2">
<h2 class="anchored" data-anchor-id="quelle-méthode-choisir">Quelle méthode choisir ?</h2>
<p>On va se servir de la matrice de distance d’origine pour tester l’adéquation de la méthode choisie pour reconstruire l’arbre aux données de départ.</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>data_NJ <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="fu">as.vector</span>(distance_matrix),</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> NJ_tree_rooted <span class="sc">|&gt;</span> <span class="fu">cophenetic</span>() <span class="sc">|&gt;</span> <span class="fu">as.dist</span>() <span class="sc">|&gt;</span> <span class="fu">as.vector</span>()</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a>corcoeff <span class="ot">&lt;-</span> <span class="fu">cor</span>(data_NJ<span class="sc">$</span>x, data_NJ<span class="sc">$</span>y)<span class="sc">^</span><span class="dv">2</span></span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> data_NJ, <span class="at">mapping =</span> <span class="fu">aes</span>(<span class="at">x =</span> x, <span class="at">y =</span> y)) <span class="sc">+</span></span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method =</span> <span class="st">"lm"</span>, <span class="at">color =</span> <span class="st">"red"</span>, <span class="at">size =</span> .<span class="dv">8</span>) <span class="sc">+</span></span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">annotate</span>(<span class="st">"text"</span>, <span class="at">x =</span> .<span class="dv">025</span>, <span class="at">y =</span> .<span class="dv">055</span>, <span class="at">color =</span> <span class="st">"red"</span>,</span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">label =</span> <span class="fu">paste0</span>(<span class="st">"Coefficient de correlation = "</span>, <span class="fu">as.character</span>(corcoeff))) <span class="sc">+</span></span>
<span id="cb23-13"><a href="#cb23-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"Est-ce que le NJ est approprié ?"</span>) <span class="sc">+</span></span>
<span id="cb23-14"><a href="#cb23-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"Distances par paires d'origine"</span>) <span class="sc">+</span></span>
<span id="cb23-15"><a href="#cb23-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"Distances par paires de l'arbre"</span>) <span class="sc">+</span></span>
<span id="cb23-16"><a href="#cb23-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Using `size` aesthetic for lines was deprecated in ggplot2 3.4.0.
ℹ Please use `linewidth` instead.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>`geom_smooth()` using formula = 'y ~ x'</code></pre>
</div>
<div class="cell-output-display">
<p><img src="biodiversity_tutorial_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Lire l’aide de la fonction <code>cophenetic()</code>. Quelle valeur calcule-t-elle ? Quel type d’objet renvoie-t-elle ?<br>
Que représente ce graphique ? Comment l’interpréter ?<br>
Comment est calculé le coefficient de corrélation ? (10 min)<br>
<br>
Recommençons avec l’arbre généré par UPGMA :</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>data_UPGMA <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="fu">as.vector</span>(distance_matrix),</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> UPGMA_tree <span class="sc">|&gt;</span> <span class="fu">cophenetic</span>() <span class="sc">|&gt;</span> <span class="fu">as.dist</span>() <span class="sc">|&gt;</span> <span class="fu">as.vector</span>()</span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(</span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> data_UPGMA,</span>
<span id="cb26-7"><a href="#cb26-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">mapping =</span> <span class="fu">aes</span>(<span class="at">x =</span> x, <span class="at">y =</span> y)</span>
<span id="cb26-8"><a href="#cb26-8" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span></span>
<span id="cb26-9"><a href="#cb26-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb26-10"><a href="#cb26-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method =</span> <span class="st">"lm"</span>, <span class="at">se =</span> <span class="cn">FALSE</span>, <span class="at">color =</span> <span class="st">"red"</span>) <span class="sc">+</span></span>
<span id="cb26-11"><a href="#cb26-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"Est-ce que l'UPGMA est appropriée ?"</span>) <span class="sc">+</span></span>
<span id="cb26-12"><a href="#cb26-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"Distances par paires d'origine"</span>) <span class="sc">+</span></span>
<span id="cb26-13"><a href="#cb26-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"Distances par paires de l'arbre"</span>) <span class="sc">+</span></span>
<span id="cb26-14"><a href="#cb26-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>`geom_smooth()` using formula = 'y ~ x'</code></pre>
</div>
<div class="cell-output-display">
<p><img src="biodiversity_tutorial_files/figure-html/unnamed-chunk-18-1.png" class="img-fluid" width="672"></p>
</div>
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cor</span>(data_UPGMA<span class="sc">$</span>x, data_UPGMA<span class="sc">$</span>y)<span class="sc">^</span><span class="dv">2</span></span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.7393009</code></pre>
</div>
</div>
<p>Qu’en concluez-vous sur cette méthode par rapport au Neighbour-Joining ? (5 min)</p>
</section>
<section id="bootstrap" class="level2">
<h2 class="anchored" data-anchor-id="bootstrap">Bootstrap</h2>
<p>Rappeler le principe du Bootstrap.<br>
À quoi sert-il dans ce contexte ? (5 min)</p>
<section id="calcul-des-valeurs-de-bootstrap" class="level3">
<h3 class="anchored" data-anchor-id="calcul-des-valeurs-de-bootstrap">Calcul des valeurs de Bootstrap</h3>
<p>Nous allons générer les indices de bootstrap pour chaque nœud de l’arbre.<br>
Commenter le bloc de code suivant :</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>bootstrap_rooted_NJ <span class="ot">&lt;-</span> <span class="fu">boot.phylo</span>(</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>  NJ_tree_rooted,</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>  alignment_BIN,</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>  \(x) {x <span class="sc">|&gt;</span></span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>      <span class="fu">dist.dna</span>(<span class="at">model =</span> <span class="st">"TN93"</span>) <span class="sc">|&gt;</span></span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a>      <span class="fu">nj</span>() <span class="sc">|&gt;</span> </span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a>      <span class="fu">root</span>(<span class="at">outgroup =</span> <span class="st">"CY012128"</span>)</span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb30-9"><a href="#cb30-9" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>
Running bootstraps:       100 / 100
Calculating bootstrap values... done.</code></pre>
</div>
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>NJ_tree_rooted<span class="sc">$</span>node.label <span class="ot">&lt;-</span> bootstrap_rooted_NJ</span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="impression-de-larbre-bootstrappé" class="level3">
<h3 class="anchored" data-anchor-id="impression-de-larbre-bootstrappé">Impression de l’arbre bootstrappé</h3>
<p>Nous allons afficher l’arbre précédent avec les valeurs de bootstrap que nous venons de calculer :</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">ignore.negative.edge=</span><span class="cn">TRUE</span>)</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggtree</span>(NJ_tree_rooted) <span class="sc">%&lt;+%</span> annotations <span class="sc">+</span></span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_tiplab</span>(<span class="fu">aes</span>(<span class="at">color =</span> year)) <span class="sc">+</span></span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_viridis</span>(<span class="at">option =</span> <span class="st">"turbo"</span>) <span class="sc">+</span></span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_label2</span>(</span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(</span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a>      <span class="at">subset =</span> <span class="sc">!</span>isTip,</span>
<span id="cb33-8"><a href="#cb33-8" aria-hidden="true" tabindex="-1"></a>      <span class="at">label =</span> label,</span>
<span id="cb33-9"><a href="#cb33-9" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb33-10"><a href="#cb33-10" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb33-11"><a href="#cb33-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"NJ raciné avec Bootstrap"</span>) <span class="sc">+</span></span>
<span id="cb33-12"><a href="#cb33-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"right"</span>) <span class="sc">+</span></span>
<span id="cb33-13"><a href="#cb33-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_rootedge</span>(<span class="at">rootedge =</span> .<span class="dv">01</span>) <span class="sc">+</span></span>
<span id="cb33-14"><a href="#cb33-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_tree2</span>()</span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="biodiversity_tutorial_files/figure-html/unnamed-chunk-20-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Commenter cet arbre. Comment améliorer cet arbre en considérant les valeurs de bootstrap ? (5 min)</p>
</section>
<section id="on-retire-les-noeuds-faibles" class="level3">
<h3 class="anchored" data-anchor-id="on-retire-les-noeuds-faibles">On retire les noeuds faibles</h3>
<p>Commenter le bloc de code suivant :</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>NJ_tree_rooted_collapsed <span class="ot">&lt;-</span> NJ_tree_rooted</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>toCollapse <span class="ot">&lt;-</span> <span class="fu">match</span>(</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">which</span>(bootstrap_rooted_NJ <span class="sc">&lt;</span> <span class="dv">70</span>) <span class="sc">+</span> <span class="fu">length</span>(NJ_tree_rooted<span class="sc">$</span>tip.label),</span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a> NJ_tree_rooted<span class="sc">$</span>edge[, <span class="dv">2</span>]</span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a>NJ_tree_rooted_collapsed<span class="sc">$</span>edge.length[toCollapse] <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a>NJ_tree_rooted_collapsed <span class="ot">&lt;-</span> <span class="fu">di2multi</span>(NJ_tree_rooted_collapsed, <span class="at">tol =</span> <span class="fl">0.00001</span>)</span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Affichons notre nouvel arbre :</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggtree</span>(NJ_tree_rooted_collapsed) <span class="sc">%&lt;+%</span> annotations <span class="sc">+</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_tiplab</span>(<span class="fu">aes</span>(<span class="at">color =</span> year)) <span class="sc">+</span></span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_viridis</span>(<span class="at">option =</span> <span class="st">"turbo"</span>) <span class="sc">+</span></span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_label2</span>(</span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(</span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a>      <span class="at">subset =</span> <span class="sc">!</span>isTip,</span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a>      <span class="at">label =</span> label,</span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb35-10"><a href="#cb35-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"NJ raciné avec Bootstrap et sans les noeuds faibles"</span>) <span class="sc">+</span></span>
<span id="cb35-11"><a href="#cb35-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"right"</span>) <span class="sc">+</span></span>
<span id="cb35-12"><a href="#cb35-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_rootedge</span>(<span class="at">rootedge =</span> .<span class="dv">01</span>) <span class="sc">+</span></span>
<span id="cb35-13"><a href="#cb35-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_tree2</span>()</span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="biodiversity_tutorial_files/figure-html/unnamed-chunk-22-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Le comparer au précédent, comment a-t-il été modifié ? (5 min)</p>
</section>
</section>
</section>
<section id="maximum-de-parcimonie" class="level1">
<h1>Maximum de parcimonie</h1>
<p>Rappeler le principe du maximum de parcimonie.<br>
Lire l’aide de la fonction <code>optim.parsimony()</code>, à quoi correspondent les arguments <code>tree</code> et <code>data</code> ? Comment le relier au principe de la méthode ? Comment l’arbre optimal est-il obtenu ? (5 min)</p>
<section id="conversion-de-lalignement-dans-le-format-phydat" class="level2">
<h2 class="anchored" data-anchor-id="conversion-de-lalignement-dans-le-format-phydat">Conversion de l’alignement dans le format phyDat</h2>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="co"># alignment_phyDat &lt;- read.phyDat(</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a><span class="co">#   file = here("output", "alignment.fa"),</span></span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a><span class="co">#   format = "fasta"</span></span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a><span class="co"># )</span></span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a>alignment_phyDat <span class="ot">&lt;-</span> alignment <span class="sc">|&gt;</span> <span class="fu">msaConvert</span>(<span class="at">type =</span> <span class="st">"phangorn::phyDat"</span>)</span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(alignment_phyDat)</span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "phyDat"</code></pre>
</div>
</div>
</section>
<section id="génération-de-larbre-initial" class="level2">
<h2 class="anchored" data-anchor-id="génération-de-larbre-initial">Génération de l’arbre initial</h2>
<p>Générons un arbre par Neighbour-Joining qui permettra d’initialiser l’algorithme de maximum de parcimonie.</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>MP_tree_init <span class="ot">&lt;-</span> alignment_BIN <span class="sc">|&gt;</span> <span class="fu">dist.dna</span>(<span class="at">model =</span> <span class="st">"TN93"</span>) <span class="sc">|&gt;</span> <span class="fu">NJ</span>()</span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="génération-de-larbre-le-plus-parcimonieux" class="level2">
<h2 class="anchored" data-anchor-id="génération-de-larbre-le-plus-parcimonieux">Génération de l’arbre le plus parcimonieux</h2>
<p>Utiliser la fonction <code>parcimony()</code> pour calculer le score de parcimonie de l’arbre initial.</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="fu">parsimony</span>(MP_tree_init, alignment_phyDat)</span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 422</code></pre>
</div>
</div>
<p>Utiliser la fonction <code>optim.parsimony()</code> pour générer l’arbre le plus parcimonieux.</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>MP_tree <span class="ot">&lt;-</span> <span class="fu">optim.parsimony</span>(MP_tree_init, alignment_phyDat)</span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Final p-score 420 after  2 nni operations </code></pre>
</div>
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(MP_tree_init)</span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>MP_tree</span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>
Phylogenetic tree with 80 tips and 76 internal nodes.

Tip labels:
  CY013200, CY013781, CY012128, CY013613, CY012160, CY012272, ...

Unrooted; no branch lengths.</code></pre>
</div>
</div>
<p>Comparer les scores de parcimonie des deux arbres.</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggtree</span>(MP_tree, <span class="at">layout =</span> <span class="st">"circular"</span>) <span class="sc">%&lt;+%</span> annotations <span class="sc">+</span></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_tiplab</span>(<span class="fu">aes</span>(<span class="at">color =</span> year)) <span class="sc">+</span></span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_viridis</span>(<span class="at">option =</span> <span class="st">"turbo"</span>) <span class="sc">+</span></span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"Arbre maximum de parcimonie"</span>) <span class="sc">+</span></span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"right"</span>)</span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="biodiversity_tutorial_files/figure-html/unnamed-chunk-27-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Commenter l’arbre généré.<br>
Comment expliquer la similarité avec les arbres précédents ? (10 min)</p>
</section>
</section>
<section id="maximum-de-vraisemblance" class="level1">
<h1>Maximum de vraisemblance</h1>
<p>Rappeler le principe du maximum de vraisemblance. (5 min)</p>
<section id="initialisation" class="level2">
<h2 class="anchored" data-anchor-id="initialisation">Initialisation</h2>
<p>Similairement à la méthode du maximum de parcimonie, la méthode doit être initialisée avec un arbre :</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>ML_tree_init <span class="ot">&lt;-</span> alignment_BIN <span class="sc">|&gt;</span> <span class="fu">dist.dna</span>(<span class="at">model =</span> <span class="st">"TN93"</span>) <span class="sc">|&gt;</span> <span class="fu">NJ</span>()</span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<section id="calcul-de-la-vraisemblance-de-larbre-initial" class="level3">
<h3 class="anchored" data-anchor-id="calcul-de-la-vraisemblance-de-larbre-initial">Calcul de la vraisemblance de l’arbre initial</h3>
<p>Utiliser la fonction <code>pml()</code> pour calculer la vraisemblance de l’arbre initial avec le paramètre <code>k = 4</code>. Ce paramètre décrit le nombre de classes de vitesses évolutives utilisées dans le modèle, la valeur <code>k = 4</code> est une valeur par défaut courante.</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="fu">pml</span>(ML_tree_init, alignment_phyDat, <span class="at">k =</span> <span class="dv">4</span>)</span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>model: JC+G(4) 
loglikelihood: -5641.785 
unconstrained loglikelihood: -4736.539 
Discrete gamma model
Number of rate categories: 4 
Shape parameter: 1 

Rate matrix:
  a c g t
a 0 1 1 1
c 1 0 1 1
g 1 1 0 1
t 1 1 1 0

Base frequencies:  
   a    c    g    t 
0.25 0.25 0.25 0.25 </code></pre>
</div>
</div>
<p>La fonction a bien tourné mais elle est sensible à la nature des données, on va donc vérifier que le jeu de données est bien propre :</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>alignment_phyDat <span class="sc">|&gt;</span> <span class="fu">as.character</span>() <span class="sc">|&gt;</span> <span class="fu">table</span>()</span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>
    -     a     c     g     k     m     r     s     t     w 
  147 45595 27170 30613     1     2     1     1 32549     1 </code></pre>
</div>
</div>
<p>Certains caractères de nos séquences ne correspondent pas aux bases azotées traditionnelles !<br>
Lire <a href="https://www.dnabaser.com/articles/IUPAC%20ambiguity%20codes.html">cette page</a> et identifier à quoi correspondent les lettres différentes de a, t, c et g.&nbsp;À quoi correspondent les <code>'-'</code> ? (5 min).<br>
</p>
<p>Enregistrons leur position dans un objet :</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a>na.posi <span class="ot">&lt;-</span> <span class="fu">which</span>(</span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">apply</span>(</span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.character</span>(alignment_BIN),</span>
<span id="cb51-4"><a href="#cb51-4" aria-hidden="true" tabindex="-1"></a>    <span class="dv">2</span>,</span>
<span id="cb51-5"><a href="#cb51-5" aria-hidden="true" tabindex="-1"></a>    \(x) <span class="fu">any</span>(<span class="sc">!</span>x <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"a"</span>,<span class="st">"t"</span>,<span class="st">"g"</span>,<span class="st">"c"</span>))</span>
<span id="cb51-6"><a href="#cb51-6" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb51-7"><a href="#cb51-7" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Voyons la distribution de ces lettres :<br>
Commenter le bloc ce code ci-après. (5 min)</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="fu">apply</span>(</span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.character</span>(alignment_BIN),</span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a>  <span class="dv">2</span>,</span>
<span id="cb52-4"><a href="#cb52-4" aria-hidden="true" tabindex="-1"></a>  \(x) <span class="fu">sum</span>(<span class="sc">!</span>x <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"a"</span>,<span class="st">"t"</span>,<span class="st">"g"</span>,<span class="st">"c"</span>))</span>
<span id="cb52-5"><a href="#cb52-5" aria-hidden="true" tabindex="-1"></a>) <span class="sc">|&gt;</span></span>
<span id="cb52-6"><a href="#cb52-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot</span>(</span>
<span id="cb52-7"><a href="#cb52-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">type =</span> <span class="st">"l"</span>,</span>
<span id="cb52-8"><a href="#cb52-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">col  =</span> <span class="st">"blue"</span>,</span>
<span id="cb52-9"><a href="#cb52-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">xlab =</span> <span class="st">"Position in HA segment"</span>,</span>
<span id="cb52-10"><a href="#cb52-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">ylab =</span> <span class="st">"Number of NAs"</span></span>
<span id="cb52-11"><a href="#cb52-11" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="biodiversity_tutorial_files/figure-html/unnamed-chunk-32-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Retirons les lettres différentes de a, t, c et g et vérifions que ça a bien fonctionné :</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a>alignment_cropped <span class="ot">&lt;-</span> alignment_BIN[, <span class="sc">-</span>na.posi]</span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a>alignment_cropped <span class="sc">|&gt;</span> <span class="fu">as.character</span>() <span class="sc">|&gt;</span> <span class="fu">table</span>()</span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>
    a     c     g     t 
43402 25104 28828 30346 </code></pre>
</div>
</div>
<p>Reconvertissons le jeu de donné trié au bon format :</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a>alignment_cropped_phyDat <span class="ot">&lt;-</span> <span class="fu">as.phyDat</span>(alignment_cropped)</span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Regénérer l’arbre d’initialisation avec notre jeu de donnée nettoyé. (5 min)</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a>ML_tree_init <span class="ot">&lt;-</span> alignment_cropped <span class="sc">|&gt;</span> <span class="fu">dist.dna</span>(<span class="at">model =</span> <span class="st">"TN93"</span>) <span class="sc">|&gt;</span> <span class="fu">NJ</span>()</span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a>ML_tree_init_L <span class="ot">&lt;-</span> <span class="fu">pml</span>(ML_tree_init, alignment_cropped_phyDat, <span class="at">k =</span> <span class="dv">4</span>)</span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a>ML_tree_init_L</span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>model: JC+G(4) 
loglikelihood: -5184.119 
unconstrained loglikelihood: -4043.367 
Discrete gamma model
Number of rate categories: 4 
Shape parameter: 1 

Rate matrix:
  a c g t
a 0 1 1 1
c 1 0 1 1
g 1 1 0 1
t 1 1 1 0

Base frequencies:  
   a    c    g    t 
0.25 0.25 0.25 0.25 </code></pre>
</div>
</div>
</section>
</section>
<section id="génération-de-larbre-le-plus-vraisemblable" class="level2">
<h2 class="anchored" data-anchor-id="génération-de-larbre-le-plus-vraisemblable">Génération de l’arbre le plus vraisemblable</h2>
<p>Utiliser la fonction <code>optim.pml()</code> pour générer l’arbre optimal avec les paramètres <code>model = "GTR", inv = .2,optNni = TRUE, optBf = TRUE, optQ = TRUE, optGamma = TRUE</code>, ce qui correspond à un modèle GTR + G(4) + I (Generalised time reversible + Gamma(4) + Invariants) très standard pour ce genre d’application. Il est également possible d’utiliser la fonction <code>modelTest()</code> pour choisir un modèle approprié automatiquement.</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a>ML_tree <span class="ot">&lt;-</span> <span class="fu">optim.pml</span>(</span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a>  ML_tree_init_L,</span>
<span id="cb58-3"><a href="#cb58-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">model    =</span> <span class="st">"GTR"</span>,</span>
<span id="cb58-4"><a href="#cb58-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">inv      =</span> .<span class="dv">2</span>,</span>
<span id="cb58-5"><a href="#cb58-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">optNni   =</span> <span class="cn">TRUE</span>,</span>
<span id="cb58-6"><a href="#cb58-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">optBf    =</span> <span class="cn">TRUE</span>,</span>
<span id="cb58-7"><a href="#cb58-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">optQ     =</span> <span class="cn">TRUE</span>,</span>
<span id="cb58-8"><a href="#cb58-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">optGamma =</span> <span class="cn">TRUE</span></span>
<span id="cb58-9"><a href="#cb58-9" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>optimize edge weights:  -5139.288 --&gt; -5121.539 
optimize base frequencies:  -5121.539 --&gt; -5121.31 
optimize rate matrix:  -5121.31 --&gt; -4933.863 
optimize shape parameter:  -4933.863 --&gt; -4919.639 
optimize edge weights:  -4919.639 --&gt; -4919.323 
optimize topology:  -4919.323 --&gt; -4916.185  NNI moves:  2 
optimize base frequencies:  -4916.185 --&gt; -4915.89 
optimize rate matrix:  -4915.89 --&gt; -4915.868 
optimize shape parameter:  -4915.868 --&gt; -4915.868 
optimize edge weights:  -4915.868 --&gt; -4915.867 
optimize topology:  -4915.867 --&gt; -4915.867  NNI moves:  0 
optimize base frequencies:  -4915.867 --&gt; -4915.866 
optimize rate matrix:  -4915.866 --&gt; -4915.866 
optimize shape parameter:  -4915.866 --&gt; -4915.866 
optimize edge weights:  -4915.866 --&gt; -4915.866 
optimize base frequencies:  -4915.866 --&gt; -4915.866 
optimize rate matrix:  -4915.866 --&gt; -4915.866 
optimize shape parameter:  -4915.866 --&gt; -4915.866 
optimize edge weights:  -4915.866 --&gt; -4915.866 </code></pre>
</div>
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a>ML_tree</span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>model: GTR+G(4) 
loglikelihood: -4915.866 
unconstrained loglikelihood: -4043.367 
Discrete gamma model
Number of rate categories: 4 
Shape parameter: 0.2829869 

Rate matrix:
          a          c         g          t
a 0.0000000  2.3834491 8.2951462  0.8556857
c 2.3834491  0.0000000 0.1486637 10.0773098
g 8.2951462  0.1486637 0.0000000  1.0000000
t 0.8556857 10.0773098 1.0000000  0.0000000

Base frequencies:  
        a         c         g         t 
0.3416376 0.1953256 0.2243397 0.2386970 </code></pre>
</div>
</div>
<p>Comparons notre arbre optimal avec l’arbre initial généré par Neighbour-Joining :</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="fu">anova</span>(ML_tree_init_L, ML_tree)</span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Likelihood Ratio Test Table
  Log lik.  Df Df change Diff log lik. Pr(&gt;|Chi|)    
1  -5184.1 158                                       
2  -4915.9 166         8        536.51  &lt; 2.2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="fu">AIC</span>(ML_tree_init_L, ML_tree)</span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>                df      AIC
ML_tree_init_L 158 10684.24
ML_tree        166 10163.73</code></pre>
</div>
</div>
<p>Lire l’aide de la fonction <code>AIC</code>. À quoi sert l’indice d’Akaike ? (5 min) Affichons notre super arbre !</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a>ML_tree_rooted <span class="ot">&lt;-</span> ML_tree<span class="sc">$</span>tree <span class="sc">|&gt;</span> <span class="fu">root</span>(<span class="at">outgroup =</span> <span class="st">"CY012128"</span>) <span class="sc">|&gt;</span> <span class="fu">ladderize</span>()</span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggtree</span>(ML_tree_rooted) <span class="sc">%&lt;+%</span> annotations <span class="sc">+</span></span>
<span id="cb66-3"><a href="#cb66-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_tiplab</span>(<span class="fu">aes</span>(<span class="at">color =</span> year)) <span class="sc">+</span></span>
<span id="cb66-4"><a href="#cb66-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_viridis</span>(<span class="at">option =</span> <span class="st">"turbo"</span>) <span class="sc">+</span></span>
<span id="cb66-5"><a href="#cb66-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_label2</span>(</span>
<span id="cb66-6"><a href="#cb66-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(</span>
<span id="cb66-7"><a href="#cb66-7" aria-hidden="true" tabindex="-1"></a>      <span class="at">subset =</span> <span class="sc">!</span>isTip,</span>
<span id="cb66-8"><a href="#cb66-8" aria-hidden="true" tabindex="-1"></a>      <span class="at">label =</span> label,</span>
<span id="cb66-9"><a href="#cb66-9" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb66-10"><a href="#cb66-10" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb66-11"><a href="#cb66-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"Maximum de vraisemblance"</span>) <span class="sc">+</span></span>
<span id="cb66-12"><a href="#cb66-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"right"</span>) <span class="sc">+</span></span>
<span id="cb66-13"><a href="#cb66-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_rootedge</span>(<span class="at">rootedge =</span> .<span class="dv">01</span>) <span class="sc">+</span></span>
<span id="cb66-14"><a href="#cb66-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_tree2</span>()</span></code><button title="Copier vers le presse-papier" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="biodiversity_tutorial_files/figure-html/unnamed-chunk-38-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copié");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copié");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
          // target, if specified
          link.setAttribute("target", "_blank");
          // default icon
          link.classList.add("external");
      }
    }
});
</script>
</div> <!-- /content -->



</body></html>